crates/net/downloaders/src/test_utils/file_client.rs

```diff
@@ -62,19 +69,82 @@ impl FileClient {
     /// Create a new file client from a file path.
     pub async fn new<P: AsRef<Path>>(path: P) -> Result<Self, FileClientError> {
         let file = File::open(path).await?;
-        FileClient::from_file(file)
+        FileClient::from_file(file).await
     }
 
     /// Initialize the [`FileClient`](FileClient) with a file directly.
-    pub(crate) fn from_file(file: File) -> Result<Self, FileClientError> {
-        let reader = BufReader::new(file);
-
-        Ok(Self {
-            reader,
-            headers: HashMap::new(),
-            bodies: HashMap::new(),
-            is_syncing: Arc::new(Default::default()),
-        })
+    pub(crate) async fn from_file(file: File) -> Result<Self, FileClientError> {
+        // get file len from metadata before reading
+        let metadata = file.metadata().await?;
+        let file_len = metadata.len();
+
+        let mut reader = BufReader::new(file);
+
+        let mut headers = HashMap::new();
+        let mut hash_to_number = HashMap::new();
+        let mut bodies = HashMap::new();
+
+        // TODO: replace with builtin const fn for rlp headers?
+        // max rlp header length is 9
+        let max_header_len = 9;
+
+        let mut block_num = 0;
+        let mut cursor = 0u64;
+        while cursor < file_len {
+            // need to avoid eof when reading
+            let header_buf_len = std::cmp::min(max_header_len, file_len - cursor);
+            let mut block_buf = vec![0u8; header_buf_len as usize];
+            reader.read_exact(&mut block_buf).await?;
+
+            let block_buf_copy = block_buf.clone();
+
+            // TODO: change rlp header decode? could just read the entire thing into memory and not
+            // use a buffered reader...
+            //
+            // # SAFETY
+            // We know that we are reading a rlp header, but the rlp decoding will check the length
+            // of the buffer after decoding the payload length, to make sure the buffer is large
+            // enough to contain the object being decoded.
+            //
+            // We do not require this check, as it is done again when we decode the RawBlockBody.
+            // This is done to ensure that we can get the exact payload length of the block, so we
```

Comment:
decode does advance the buffer so we can always get this by comparing the lengths